### PR TITLE
add BitBasicList component (#269 )

### LINKED ISSUE #269: Add Basic List Component
Add **Basic List** Component based on [Microsoft FluentUI Design](https://developer.microsoft.com/en-us/fluentui#/controls/web/list).

### PR DESCRIPTION
**Pull request by zoha-shobbar**
This closes #269

### CODE REVIEW

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+using static Bit.Client.Web.BlazorUI.Tests.Lists.BitBasicListTest;
+
+namespace Bit.Client.Web.BlazorUI.Tests.Lists
+{
+    [TestClass]
+    public class BitBasicListTests : BunitTestContext
+    {
+
+        [DataTestMethod, DataRow(1000)]
+        public async Task BitBasicListShoudReturnExpectedChildElements(int itemCount)
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs:14]: The method name could contain "Render" instead of "Return"

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        {
+            var component = RenderComponent<BitBasicListTest>();
+
+            var bitList = component.Find(".bit-basic-list");
+            int comItemCount = bitList.ChildElementCount;
+
+            Assert.AreEqual(comItemCount, itemCount + 1);
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs:21]: why itemCount+1 ?

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        }
+        [DataTestMethod, DataRow(true)]
+        public async Task BitBasicListCheckWithVirtualize(bool vitualize)
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs:24]: the method name does not clearly show the test assertion

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        {
+            var component = RenderComponent<BitBasicListTest>();
+            var bitList = component.Find(".bit-basic-list");
+
+            var height = bitList.FirstElementChild.GetAttribute("style").Split(":")[1].Replace("px;", "");
+
+            Assert.AreEqual("0", height.Trim());
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs:31]: this method could check that only a subset of all items has been rendered.

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        [DataTestMethod, DataRow(false, "testItem")]
+        public async Task BitBasicListCheckWithoutVirtualize(bool vitualize, string childClass)
+        {
+            var component = RenderComponent<BitBasicListTest>(parameters => parameters.Add(p => p.Virtualize, vitualize));
+            var bitList = component.Find(".bit-basic-list");
+
+            bool vitualizeIsActive = true;
+            var itemCalss = bitList.FirstElementChild.FirstElementChild.GetAttribute("class");
+            if (itemCalss == childClass && bitList.FirstElementChild.GetAttribute("style") == null && bitList.FirstElementChild.GetAttribute("class") == null)
+                vitualizeIsActive = false;
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs:44]: always use { }

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+            var bitList = component.Find(".bit-basic-list");
+
+            var height = bitList.FirstElementChild.GetAttribute("style").Split(":")[1].Replace("px;", "");
+
+            Assert.AreEqual("0", height.Trim());
+        }
+
+
+        [DataTestMethod, DataRow(false, "testItem")]
+        public async Task BitBasicListCheckWithoutVirtualize(bool vitualize, string childClass)
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs:36]: If it's a check without "Virtualize" then there is no need to take the option from method arguments. it is false by definition in this method

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        {
+            var component = RenderComponent<BitBasicListTest>(parameters => parameters.Add(p => p.Virtualize, vitualize));
+            var bitList = component.Find(".bit-basic-list");
+
+            bool vitualizeIsActive = true;
+            var itemCalss = bitList.FirstElementChild.FirstElementChild.GetAttribute("class");
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs:42]: typo: itemCalss  =>  itemClass

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+            if (itemCalss == childClass && bitList.FirstElementChild.GetAttribute("style") == null && bitList.FirstElementChild.GetAttribute("class") == null)
+                vitualizeIsActive = false;
+
+            Assert.AreEqual(vitualize, vitualizeIsActive);
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs:46]: I can't understand this test, please explain it to me later

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+        [DataTestMethod, DataRow("RoleItem")]
+        public async Task BitBasicListCheckForCorrectRole(string role)
+        {
+            var component = RenderComponent<BitBasicListTest>(parameters => parameters.Add(p => p.Role, role));
+            var bitList = component.Find(".bit-basic-list");
+
+            var itemRole = bitList.GetAttribute("role");
+            
+            Assert.AreEqual(role,itemRole);
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs:57]: code formatting issue

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/Lists/BitBasicList.razor [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+                    <p class="text-secondary">
+                        @LoadingText
+                    </p>
+                    }
+                    else
+                    {
+                    <p>
+                        Loading&hellip;
+                    </p>
+                    }
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Components/Lists/BitBasicList.razor:33]: code formatting issue

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/Lists/BitBasicList.razor [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+            </Placeholder>
+        </Virtualize>}
+    else
+    {
+        @foreach (var item in Items ?? Array.Empty<TItem>())
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Components/Lists/BitBasicList.razor:39]: The "Items" property should not be null, so this null check operator (??) is redundant

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/Lists/BitBasicList.razor.cs [hunk 12]
```diff
@@ -0,0 +1,56 @@
+﻿using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components;
+
+namespace Bit.Client.Web.BlazorUI
+{
+    public partial class BitBasicList<TItem>
+    {
+        [Parameter] public ICollection<TItem> Items { get; set; }
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Components/Lists/BitBasicList.razor.cs:9]: initialize this parameter with an empty collection

File: src/Client/Web/Playground/Bit.Client.Web.BlazorUI.Playground/Web/Pages/Component.razor [hunk 13]
```diff
@@ -417,4 +417,18 @@
...
+        <BitBasicList Items="people" LoadingText="please wait data is loading..." Virtualize="true" ItemPerPage="20">
+            <RowTemplate Context="person">
+                <div style=" border: 2px #8a8886 solid; border-radius: 10px; padding: 5px 20px; margin: 10px;">
+                    <p>Id : <strong >@person.Id</strong></p>
+                    <p>Full Name :<strong>@person.FirstName @person.LastName</strong>  </p>
+                    <p>Job: <strong>@person.Job</strong></p>
+                </div>                    
+            </RowTemplate>
+        </BitBasicList>
+    </div>
```
Comment (msynk) [inline src/Client/Web/Playground/Bit.Client.Web.BlazorUI.Playground/Web/Pages/Component.razor:433]: please provide some more examples in order to explain other features of the BitBasicList

**Review by msynk — CHANGES_REQUESTED** [review]
please resolve the Nullable Reference Type warnings in the files changed/added in this PR

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs (hunk 2 above)
Comment (zoha-shobbar) [reply src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs]: there was a mistake and should be +2,
When we use virtualization, two additional parts are created, that's why this method use itemCount +2

**Review by msynk — CHANGES_REQUESTED** [review]

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+
+            var bitList = component.Find(".bit-bsc-lst");
+
+            var visualClass = visual == Visual.Cupertino ? "cupertino" : visual == Visual.Material ? "material" : "fluent";
+            Assert.IsTrue(bitList.ClassList.Contains($"bit-bsc-lst-{visualClass}"));
+
+            int comItemCount = bitList.ChildElementCount;
+            if (virtualize)
+            {
+                Assert.AreEqual(comItemCount, itemCount + 2);
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs:43]: this assertion is not correct, in virtualized mode, the number of elements rendered is about the size of a single page not about the size of all of the data

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+
+            var bitList = component.Find(".bit-bsc-lst");
+
+            var visualClass = visual == Visual.Cupertino ? "cupertino" : visual == Visual.Material ? "material" : "fluent";
+            Assert.IsTrue(bitList.ClassList.Contains($"bit-bsc-lst-{visualClass}"));
+
+            int comItemCount = bitList.ChildElementCount;
+            if (virtualize)
+            {
+                Assert.AreEqual(comItemCount, itemCount + (overScanCount + 2));
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs:80]: again, in virtualized mode, the number of elements rendered is about the size of a single page not about the size of all of the data

Comment (msynk) [reply src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs:80]: and one more thing, the default value of OverScanCount in the Blazor Virtualize component is 3, so this value is should be considered in the other tests.

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+            {
+                parameters.Add(p => p.Visual, visual);
+                parameters.Add(p => p.Virtualize, virtualize);
+                parameters.Add(p => p.Items, GetTestData(itemCount));
+            });
+
+            var bitList = component.Find(".bit-bsc-lst");
+
+            var visualClass = visual == Visual.Cupertino ? "cupertino" : visual == Visual.Material ? "material" : "fluent";
+            Assert.IsTrue(bitList.ClassList.Contains($"bit-bsc-lst-{visualClass}"));
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs:38]: this assertion is not related to this test. please move it to its own test describing the behavior of classes of the component

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+                parameters.Add(p => p.Visual, visual);
+                parameters.Add(p => p.Virtualize, virtualize);
+                parameters.Add(p => p.Items, GetTestData(itemCount));
+                parameters.Add(p => p.OverscanCount, overScanCount);
+            });
+
+            var bitList = component.Find(".bit-bsc-lst");
+
+            var visualClass = visual == Visual.Cupertino ? "cupertino" : visual == Visual.Material ? "material" : "fluent";
+            Assert.IsTrue(bitList.ClassList.Contains($"bit-bsc-lst-{visualClass}"));
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs:75]: again, this assertion is not related to this test. please move it to its own test describing the behavior of classes of the component

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+            {
+                parameters.Add(p => p.Visual, visual);
+                parameters.Add(p => p.Items, GetTestData(itemCount));
+                parameters.Add(p => p.Role, role);
+                parameters.Add(p => p.Virtualize, virtualize);
+            });
+            var bitList = component.Find(".bit-bsc-lst");
+
+            var visualClass = visual == Visual.Cupertino ? "cupertino" : visual == Visual.Material ? "material" : "fluent";
+            Assert.IsTrue(bitList.ClassList.Contains($"bit-bsc-lst-{visualClass}"));
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Lists/BitBasicListTests.cs:110]: this assertion is not related to this test. please move it to its own test describing the behavior of classes of the component

File: src/Client/Web/Bit.Client.Web.BlazorUI/Scripts/bit.client.web.ts [hunk 19]
```diff
@@ -1,5 +1,6 @@
-﻿class Bit {
-    static setProperty(element: { [key: string]: any },property: string,value: any): void {
+﻿
+class Bit {
+    static setProperty(element: { [key: string]: any }, property: string, value: any): void {
         element[property] = value;
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Scripts/bit.client.web.ts:4]: please remove these changes, which are not related to this PR

File: src/Client/Web/Playground/Bit.Client.Web.BlazorUI.Playground/Web/Pages/Component.razor.cs [hunk 20]
```diff
@@ -217,6 +217,39 @@ private void HideMessageBar(MouseEventArgs args)
             IsMessageBarHidden = true;
         }
 
-        private string RatingChangedText = "";
```
Comment (msynk) [inline src/Client/Web/Playground/Bit.Client.Web.BlazorUI.Playground/Web/Pages/Component.razor.cs:220]: why this line is removed?

File: src/Client/Web/Playground/Bit.Client.Web.BlazorUI.Playground/Web/Pages/Component.razor.scss [hunk 21]
```diff
@@ -56,3 +56,19 @@
 .pivot-container {
     margin: 20px 10px;
 }
+
+
+.lst-custom-style {
+
```
Comment (msynk) [inline src/Client/Web/Playground/Bit.Client.Web.BlazorUI.Playground/Web/Pages/Component.razor.scss:62]: redundant line?

**Review by msynk — CHANGES_REQUESTED** [review]

**Comment by msynk** [comment]
Since the assignee of the related issue is changed, I'm closing this PR.
